routines/pool.go

```diff
@@ -0,0 +1,138 @@
+package routines
+
+import (
+	"sync"
+)
+
+// Pool is a FIFO go-routine pool.
+type Pool struct {
+	wq        []*work
+	terminate bool
+	wqMutex   sync.Mutex // protects wq and terminate
+
+	workChan chan *work
+
+	schedulerNotifyChan chan struct{}
+
+	terminateWg sync.WaitGroup
+}
+
+type work struct {
+	fn       WorkFn
+	userData interface{}
+}
+
+// WorkFn is a function that is executed by the pool workers.
+type WorkFn func(userData interface{})
```

Comment:
I don't think we need the userData parameter. The caller can always provide data by using a _closure_ or an object, and those approaches would not degenerate the types to `interface{}`. If we do that then we don't even need the `work` struct anymore.

Reply:
That was the first version and then I stumbled about the issue that I was referencing in the closure a variable which changed outside of the closure and got very unexpected behavior, like:

```go
for _, build := range builds {
    pool.Queue(func() { fmt.Printf("build: %+v\n", build) })
}
```

build was already updated before the go routine run and all go-routines were running only for the last `build` in the `builds` slice

Do you maybe have a better idea to prevent the issue?

Reply:
I don't think accepting `interface{}` as parameter fixes the issue, because you can still pass references and change them, and you can still pass a closure, even when accepting the parameter. To me it looks like accepting the parameter makes the code more complicated and induces a weakening of the type system.

I think any approach will require the caller to make sure it don't modify anything being used on the job. We can have this without the need to add parameters.

Reply:
> I don't think accepting `interface{}` as parameter fixes the issue, because you can still pass references

It's for me non-intuitive that uploadAndRecord might be executed for `t.task` and `t.inputs` of other loop iterations in this example:
```go
for _, t := range taskToRun {
[...]
		c.uploadRoutinePool.Queue(
			func() {
				c.uploadAndRecord(ctx, t.task, t.inputs, outputs, runResult)
			})
	}
```

It's not intuitive for me because I don't pass a reference to `t` here.
It seems very easy to miss and error prone.

> and you can still pass a closure, even when accepting the parameter. 

true

> I think any approach will require the caller to make sure it don't modify anything being used on the job. We can have this without the need to add parameters.

Here is the `internal/command/run.go` code with a pool version that only accepts a function as parameter:
https://github.com/simplesurance/baur/blob/0de2dd2941707418eb81febb0aa9cecced5aacb8/internal/command/run.go#L274

Do you have another suggestion how I can pass in `t` here to prevent the described issue?

The only idea I have is to wrap the `Queue()` call  in another function to create a copy of `t`, like:
```go
func(t *pendingTask) {                                                                                                                                                                                                                                                   
       c.uploadRoutinePool.Queue(func() {                                                                                                                                                                                                     
              c.uploadAndRecord(ctx, t.task, t.inputs, outputs, runResult)                                                                                                         
       })                                                                                                                                                             
 }(t)
```

Reply:
I like you second approach. Another way would be to make sure the closure captures a variable that is declared inside the loop, effectively given each closure it's own variable with:

```go
for _, t := range taskToRun {
    t := t // make sure each each closure captures it's own variable
    [...]
}
```

See
https://play.golang.org/p/9Y-VBxT6IwR

Reply:
Thanks, I changed it as suggested.
You convinced me that the userData parameter did not made it better only more complicated :-)